text/3263-pre-release-sticky.md

```diff
@@ -0,0 +1,115 @@
+- Feature Name: `pre-release-sticky`
+- Start Date: 2022-05-10
+- RFC PR: [rust-lang/rfcs#3263](https://github.com/rust-lang/rfcs/pull/3263)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+The following assume:
+
+ - A release is a version without a "pre-release" tag (like `1.0.0`)
+ - A pre-release is a version with a "pre-release" tag (like `1.0.0-alpha`)
+
+Cargo will now apply no precedence rule on pre-release by default, this mean that `1.0.0-alpha.0` will be interpreted as `=1.0.0-alpha.0` by cargo instead of `^1.0.0-alpha.0`.
+
+# Motivation
+[motivation]: #motivation
+
+The current behavior often break pre-release build, [Semver 2.0](https://semver.org/#spec-item-11) rules for pre-release doesn't include breaking change concept between pre-release. This mean that any new pre-release would be considerate by cargo as a "compatible update". Even worse, final version also match the requierement of any pre-release. Say otherwise, if an user put `version = "1.0.0-alpha.0"` in their `Cargo.toml` this will be considered by cargo as compatible with any pre-release that follow `1.0.0-alpha.1`, `1.0.0-alpha.2`, `1.0.0-beta.0`, etc... and also `1.0.0`, `1.1.0`, etc... (But not `2.0.0`). This can also lead to security problem.
+
+This is an open problem since some time, [cargo #2222](https://github.com/rust-lang/cargo/issues/2222) but have receive some input in Oct 2021, specially in this [forum post](https://internals.rust-lang.org/t/changing-cargo-semver-compatibility-for-pre-releases/14820) where this proposition have been suggested.
+
+Finally, contrary to release version where most user benefits from having cargo pick the latest compatible version as default, there is little sense to do the same for pre-release version cause there are very likely be breaking change between all pre-release version. This is a trap that most user don't expect to happen (why does my project don't compile anymore ?!?). Pre-release version precedence of `Cargo.toml` should default to something more user friendly. User expect bugs from a pre-release but no one want their build suddenly break when a new pre-release is out.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+By default Cargo follow Semver precedence rules for release of a crate but will make no assumption for pre-release.
+
+```toml
+[dependencies.foo]
+version = "1.0.0-alpha.0"
+```
+
+will now be considered by Cargo as:
+
+```toml
+[dependencies.foo]
+version = "=1.0.0-alpha.0"
+```
+
+while release version like:
+
+```toml
+[dependencies.foo]
+version = "1.0.0"
+```
+
+are considered by Cargo as:
+
+```toml
+[dependencies.foo]
+version = "^1.0.0"
+```
+
+The latter doesn't change.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Do it transparently
```

Comment:
I think we could use crater to try this out and fallback to opt-in solution if this doesn't work.

Reply:
I think my next step will be to try implement this myself on Cargo, I think I will try the alternative solution in first cause that I believe would fix Semver. Also I believe it would be more simple to match a patch to semver crates than trying to find how to change resolver cargo behavior

Reply:
After patching semver I realize the problem is deeper than I expected, in early day of rust we try to copy npm behaviour without much questioning, but 1 our copy of npm is not perfect 2 npm is wrong in many points. Specially on precedence rules, npm is in clear contradiction with Semver 2.0. They make a lot of exception for pre-release, while semver is clear about precedence rules. This lead to even wanting to add theses npm rules into Semver https://github.com/semver/semver/pull/584. This is caused by npm default to `=` so people have been using comparator operator as caret operator, so they run into problem cause pre-release was included (as Semver said) they so stop from following Semver specification and make exceptions.

The [patch](https://github.com/Stargateur/semver/commit/804bd69d577954db5b5143efeb43299f7a16f30f) I did was very informative on the implication and the root problem of the misconception about what is precedence and what are compatibility rule in Semver. The good thing is that the patch, make the implementation simpler. While change the rule I not only apply this RFC proposition but also fix the incorrect behaviour of semver implementation of precedence.

All of this make me think we should maybe stop trying to follow npm. If this situation continue I think we should stop using semver specification cause I think they will only reflect what npm do. We should not use a convention just because a lot of javascript code use it if the convention is wrong.

As always, this is just my opinion, anyone can disagree.